traffic_ops/traffic_ops_golang/ats/parentdotconfig.go

```diff
@@ -0,0 +1,1393 @@
+		}
+		if v, ok := dsParams[ParentConfigParamMaxUnavailableServerRetries]; ok {
+			ds.MSOMaxUnavailableServerRetries = v
+		} else {
+			ds.MSOMaxUnavailableServerRetries = ParentConfigDSParamDefaultMaxUnavailableServerRetries
+		}
+		dses[i] = ds
+	}
+	return dses, nil
+}
+
+func getParentConfigDSParamsRaw(tx *sql.Tx, qry string, dsNames []string) (map[tc.DeliveryServiceName]map[string]string, error) {
+	rows, err := tx.Query(qry, pq.Array(dsNames))
+	if err != nil {
+		return nil, errors.New("querying: " + err.Error())
+	}
+	defer rows.Close()
+
+	params := map[tc.DeliveryServiceName]map[string]string{}
+	for rows.Next() {
+		dsName := tc.DeliveryServiceName("")
+		pName := ""
+		pVal := ""
+		if err := rows.Scan(&dsName, &pName, &pVal); err != nil {
+			return nil, errors.New("scanning: " + err.Error())
+		}
+		log.Errorf("DEBUG getParentConfigDSParamsRaw dsName %+v pName %+v pVal %+v\n", dsName, pName, pVal)
+		if _, ok := params[dsName]; !ok {
+			params[dsName] = map[string]string{}
+		}
+		params[dsName][pName] = pVal
+	}
+	return params, nil
+}
+
+type ParentInfo struct {
+	Host            string
+	Port            int
+	Domain          string
+	Weight          string
+	UseIP           bool
+	Rank            int
+	IP              string
+	PrimaryParent   bool
+	SecondaryParent bool
+}
+
+// parentData parent_data in Perl
+func getParentInfo(tx *sql.Tx, server *ServerInfo) (map[string][]ParentInfo, error) {
+	log.Errorf("DEBUG PCGen getParentInfo\n")
+	parentInfos := map[string][]ParentInfo{}
+
+	serverDomain, ok, err := getCDNDomainByProfileID(tx, server.ProfileID)
+	if err != nil {
+		return nil, errors.New("getting CDN domain from profile ID: " + err.Error())
+	} else if !ok || serverDomain == "" {
+		return parentInfos, nil // TODO warn? Perl doesn't.
+	}
+	log.Errorf("DEBUG PCGen getParentInfo got CDN domain '%+v'\n", serverDomain)
+
+	profileCaches, deliveryServices, err := getServerParentCacheGroupProfiles(tx, server)
+	if err != nil {
+		return nil, errors.New("getting server parent cachegroup profiles: " + err.Error())
+	}
+
+	log.Errorf("DEBUG PCGen getParentInfo got profileCaches len '%+v'\n", len(profileCaches))
+	log.Errorf("DEBUG PCGen getParentInfo got deliveryServices len '%+v'\n", len(deliveryServices))
+
+	// log.Errorf("DEBUG PCGen getParentInfo got profileCaches '%++v'\n", profileCaches)
+	// log.Errorf("DEBUG PCGen getParentInfo got deliveryServices '%++v'\n", deliveryServices)
+
+	// note deliveryServies also contains an "all" key
+	// originFQDN is "prefix" in Perl; ds is not really a "ds", that's what it's named in Perl
+	for originFQDN, dses := range deliveryServices {
+		for _, row := range dses {
+			profile := profileCaches[row.ProfileID]
+			if profile.NotAParent {
+				log.Errorf("DEBUG PCGen getParentInfo row " + row.ServerHost + " not a parent, skipping\n")
+				continue
+			}
+			// Perl has this check, but we only select "deliveryServices" (servers) with the right CDN in the first place
```

Comment:
I'd prefer to keep it, until this has been tested in Production for a little while, in case there turns out to be an error, to help debug the problem.